osu.Game/Tests/Visual/ModSandboxTestScene.cs

```diff
@@ -0,0 +1,107 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using JetBrains.Annotations;
+using osu.Game.Beatmaps;
+using osu.Game.Replays;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Scoring;
+using osu.Game.Screens.Play;
+
+namespace osu.Game.Tests.Visual
+{
+    public abstract class ModSandboxTestScene : PlayerTestScene
+    {
+        protected sealed override bool HasCustomSteps => true;
+
+        public override IReadOnlyList<Type> RequiredTypes => new[]
+        {
+            typeof(ModSandboxTestScene)
+        };
+
+        protected ModSandboxTestScene(Ruleset ruleset)
+            : base(ruleset)
+        {
+        }
+
+        private ModTestCaseData currentTest;
+
+        protected void CreateModTest(ModTestCaseData testCaseData) => CreateTest(() =>
+        {
+            AddStep("set test data", () => currentTest = testCaseData);
+        });
+
+        public override void TearDownSteps()
+        {
+            AddUntilStep("test passed", () => currentTest?.PassCondition?.Invoke() ?? true);
```

Comment:
Not sure we should be passing tests by default. Personally I'd do the opposite (fail by default and only pass if explicitly overridden by the test writer), or even throw if a pass condition wasn't specified.

Reply:
Changed it to `false` by default. I think this is the better of the two since an endless + failing step will allow the user to at least test their mods further.